warp10/src/main/java/io/warp10/script/ScriptRunner.java

```diff
@@ -39,265 +68,232 @@
       this.minperiod = Long.parseLong(config.getProperty(Configuration.RUNNER_MINPERIOD));
       this.endpoint = config.getProperty(Configuration.RUNNER_ENDPOINT);
-      
-      this.executor = new ThreadPoolExecutor(1, nthreads, 30000L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>(nthreads * 256));
-  
+
+      ThreadPoolExecutor runnersExecutor = new ThreadPoolExecutor(nthreads, nthreads, 30000L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>(nthreads * 256), new NamedThreadFactory());
+      runnersExecutor.allowCoreThreadTimeOut(true);
+
+      this.executor = runnersExecutor;
+
       this.leaderLatch = null;
-      
+
       this.setDaemon(true);
       this.setName("[Warp ScriptRunner]");
-      this.start();      
+      this.start();
     }
-    
+
     if (isWorker || isScheduler) {
       extractKeys(config);
-      
+
       byte[] k = this.keystore.getKey(KeyStore.SIPHASH_KAFKA_RUNNER);
-      
+
       if (null != k) {
         this.KAFKA_MAC = SipHashInline.getKey(k);
       }
     }
-    
+
     if (isWorker) {
       for (String required: REQUIRED_PROPERTIES_WORKER) {
-        Preconditions.checkNotNull(config.getProperty(required), "Missing configuration parameter '%s'.", required);          
+        Preconditions.checkNotNull(config.getProperty(required), "Missing configuration parameter '%s'.", required);
       }
-      
+
       this.endpoint = config.getProperty(Configuration.RUNNER_ENDPOINT);
-      
+
       String zkconnect = config.getProperty(Configuration.RUNNER_KAFKA_ZKCONNECT);
       this.topic = config.getProperty(Configuration.RUNNER_KAFKA_TOPIC);
       String groupid = config.getProperty(Configuration.RUNNER_KAFKA_GROUPID);
       String clientid = config.getProperty(Configuration.RUNNER_KAFKA_CONSUMER_CLIENTID);
       String strategy = config.getProperty(Configuration.RUNNER_KAFKA_CONSUMER_PARTITION_ASSIGNMENT_STRATEGY);
       int nthreads = Integer.parseInt(config.getProperty(Configuration.RUNNER_KAFKA_NTHREADS));
       long commitPeriod = Long.parseLong(config.getProperty(Configuration.RUNNER_KAFKA_COMMITPERIOD));
-      this.nthreads = Integer.parseInt(config.getProperty(Configuration.RUNNER_NTHREADS));
-      this.executor = new ThreadPoolExecutor(1, nthreads, 30000L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>(nthreads * 256));
-      
-      this.consumerPool = new KafkaSynchronizedConsumerPool(zkconnect, topic, clientid, groupid, strategy, nthreads, commitPeriod, new ScriptRunnerConsumerFactory(this));
+
+      ThreadPoolExecutor runnersExecutor = new ThreadPoolExecutor(nthreads, nthreads, 30000L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>(nthreads * 256), new NamedThreadFactory());
+      runnersExecutor.allowCoreThreadTimeOut(true);
+
+      this.executor = runnersExecutor;
+
+      new KafkaSynchronizedConsumerPool(zkconnect, topic, clientid, groupid, strategy, nthreads, commitPeriod, new ScriptRunnerConsumerFactory(this));
```

Comment:
Code is clearer if you keep a reference to the created consumer pool

Reply:
I do agree this line looks strange without keeping a reference but not storing it in a member of the class makes it clear that it is not used/modified elsewhere.
Can a local variable do? What about the GC, could it destroy the instance if there are no ref to this instance?

Reply:
The GC issue would have to be investigated, a local variable would not work if the GC can discard the object, so for now it's probably safer to stick with the setting of the field.